libbeat/processors/timeseries/timeseries.go

```diff
@@ -0,0 +1,133 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package timeseries
+
+import (
+	"strings"
+
+	"github.com/mitchellh/hashstructure"
+
+	"github.com/elastic/beats/libbeat/asset"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+type timeseriesProcessor struct {
+	dimensions map[string]interface{}
+	prefixes   map[string]interface{}
+}
+
+// NewTimeSeriesProcessor returns a processor to add timeseries info to events
+// Events are processed to extract all their dimensions (keyword fields that
+// hold a dimension of the metrics) and compute a hash of all their values into
+// `timeseries.instance` field.
+func NewTimeSeriesProcessor(beatName string) (processors.Processor, error) {
+	fieldsYAML, err := asset.GetFields(beatName)
+	if err != nil {
+		return nil, err
+	}
+
+	fields, err := common.NewFieldsFromYAML(fieldsYAML)
+	if err != nil {
+		return nil, err
+	}
+
+	dimensions := map[string]interface{}{}
+	prefixes := map[string]interface{}{}
+	populateDimensions("", dimensions, prefixes, fields)
+
+	return &timeseriesProcessor{dimensions: dimensions, prefixes: prefixes}, nil
+}
+
+func (t *timeseriesProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	if event.TimeSeries {
+		instanceFields := common.MapStr{}
+
+		// map all dimensions & values
+		for k, v := range event.Fields.Flatten() {
+			if t.isDimension(k) {
+				instanceFields[k] = v
+			}
+		}
+
+		h, err := hashstructure.Hash(instanceFields, nil)
+		if err != nil {
+			// this should not happen, keep the event in any case
+			return event, err
+		}
+		event.Fields["timeseries"] = common.MapStr{
+			"instance": h,
+		}
+	}
+
+	return event, nil
+}
```

Comment:
I still wonder if this processor can affect CPU usage in metricbeat for metricsets with lots of keywords that can generate lots of events, like kubernetes container or system process.

Reply:
yes, maybe we ship this off by default and experimental for starters?

Reply:
nevermind, it's already off by default: https://github.com/elastic/beats/pull/10963/files#diff-8f7f0a04a8abba85477db97db6fdfa19R88

Reply:
Yeah, we can ship it as experimental and disabled by default by now, but still ship it to help developing the UI parts in the meantime.

Reply:
+1 to make it optional, but I would go to enable it by default in Metricbeat. I don't expect the overhead to be a problem on metric uses cases. Happy to do some testing before making the decission